pkg/e2e/infra/utils.go

```diff
@@ -336,3 +339,220 @@ func waitUntilAllNodesAreSchedulable(client runtimeclient.Client) error {
+			glog.Errorf("Error querying api machine %q object: %v, retrying...", targetMachine.Name, err)
+			return false, nil
+		}
+		if machine.Status.NodeRef == nil || machine.Status.NodeRef.Kind != "Node" {
+			glog.Error("Machine %q not linked to a node", machine.Name)
+			return false, nil
+		}
+
+		drainedNodeName = machine.Status.NodeRef.Name
+		node := corev1.Node{}
+
+		if err := client.Get(context.TODO(), types.NamespacedName{Name: drainedNodeName}, &node); err != nil {
+			glog.Errorf("Error querying api node %q object: %v, retrying...", drainedNodeName, err)
+			return false, nil
+		}
+
+		if !node.Spec.Unschedulable {
+			glog.Errorf("Node %q is expected to be marked as unschedulable, it is not", node.Name)
+			return false, nil
+		}
+
+		glog.Infof("Node %q is mark unschedulable as expected", node.Name)
+
+		pods := corev1.PodList{}
+		listOpt := &runtimeclient.ListOptions{}
+		listOpt.MatchingLabels(rc.Spec.Selector)
+		if err := client.List(context.TODO(), listOpt, &pods); err != nil {
+			glog.Errorf("Error querying api for Pods object: %v, retrying...", err)
+			return false, nil
+		}
+
+		podCounter := 0
+		for _, pod := range pods.Items {
+			if pod.Spec.NodeName != machine.Status.NodeRef.Name {
+				continue
+			}
+			if !pod.DeletionTimestamp.IsZero() {
+				continue
+			}
+			podCounter++
+		}
+
+		glog.Infof("Have %v pods scheduled to node %q", podCounter, machine.Status.NodeRef.Name)
+
+		// Verify we have enough pods running as well
+		rcObj := corev1.ReplicationController{}
+		key = types.NamespacedName{
+			Namespace: rc.Namespace,
+			Name:      rc.Name,
+		}
+		if err := client.Get(context.TODO(), key, &rcObj); err != nil {
+			glog.Errorf("Error querying api RC %q object: %v, retrying...", rc.Name, err)
+			return false, nil
+		}
+
+		// The point of the test is to make sure majority of the pods are rescheduled
+		// to other nodes. Pod disruption budget makes sure at most one pod
+		// owned by the RC is not Ready. So no need to test it. Though, useful to have it printed.
+		glog.Infof("RC ReadyReplicas: %v, Replicas: %v", rcObj.Status.ReadyReplicas, rcObj.Status.Replicas)
+
+		// This makes sure at most one replica is not ready
+		if rcObj.Status.Replicas-rcObj.Status.ReadyReplicas > 1 {
+			return false, fmt.Errorf("pod disruption budget not respected, node was not properly drained")
+		}
+
+		// Depends on timing though a machine can be deleted even before there is only
+		// one pod left on the node (that is being evicted).
+		if podCounter > 2 {
+			glog.Infof("Expecting at most 2 pods to be scheduled to drained node %q, got %v", machine.Status.NodeRef.Name, podCounter)
+			return false, nil
+		}
+
+		glog.Info("Expected result: all pods from the RC up to last one or two got scheduled to a different node while respecting PDB")
+		return true, nil
+	})
+
+	return drainedNodeName, err
+}
+
+func waitUntilNodeDoesNotExists(client runtimeclient.Client, nodeName string) error {
```

Comment:
there's already a lib function to check nodes are ready. We should use that and rely on ".eventually" on the ginkgo script. So we keep this library for manipulating resources atomically as free as possible from wait operations cc @ingvagabund @frobware